In [5]:
import pandas as pd
import numpy as np
from datetime import datetime, timedelta

## FFS

In [6]:
ffs_df = pd.read_csv("../static/data/sanfrancisco/ffs_temp.csv")
ffs_df

0
0     43.44
1     30.41
2     30.41
3     26.07
4     30.41
...     ...
4463  30.41
4464  30.41
4465  30.41
4466  21.72
4467  30.41

[4468 rows x 1 columns]

# 도로별 TCI

- speed(20080517~20080610).csv
    - 행: 5분 간격 시계열
        - 7488개
    - 열: SanFrancisco_edges의 행과 매핑
        - 4468개 (확인필요)
    - 값: 속도(km/h)

In [7]:
start_date = datetime.strptime("2008-05-17", "%Y-%m-%d")

In [23]:
df = pd.read_csv("../static/data/sanfrancisco/speed_san.csv").transpose().replace(0, np.nan)
df

0     1     2     3     4     5     6     7     8     9     ...  7016  \
0      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
2      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
3      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
4      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
4463   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
4464   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
4465   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
4466   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
4467   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   

      7017  7018  7019  7020  7021  7022  7023  7024  7025  
0      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
3      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
4463   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4464   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4465   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4466   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4467   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[4468 rows x 7026 columns]

In [24]:
# 날짜 단위 데이터프레임 생성
date_df = pd.DataFrame()
date_df.index = df[0].index
for i in range(0, len(df.columns), 12*24):
    cols = df.columns[i:i+12*24]
    sums = df[cols].mean(axis=1)
    colname = start_date + timedelta(days=int(i/(12*24)))

    date_df = pd.concat((date_df, sums.rename(colname)), axis=1)

date_df

2008-05-17  2008-05-18  2008-05-19  2008-05-20  2008-05-21  2008-05-22  \
0      99.990603  106.195309         NaN         NaN         NaN   93.555582   
1       8.357541         NaN   58.404092         NaN         NaN   17.943910   
2      48.824357   50.031565   86.777742   61.514348   61.934952   78.773444   
3      94.147401   66.950067   83.493598   79.771106   82.450063   81.547987   
4            NaN         NaN         NaN         NaN         NaN         NaN   
...          ...         ...         ...         ...         ...         ...   
4463   75.852678   44.747002   72.757176   60.640548   76.150534   89.068450   
4464   60.762089   53.776279   70.402195   50.081816   71.549457   57.660560   
4465   25.273887   69.427038  109.891221   75.252768   53.401202   86.004612   
4466   60.064276   54.241152   57.550148   66.515002   52.074041   59.093084   
4467         NaN         NaN   53.911431   28.725956   24.976483   54.319428   

      2008-05-23  2008-05-24  2008-05-25  2008-05-26  ...  2008-06-01  \
0      58.168927         NaN         NaN         NaN  ...         NaN   
1            NaN         NaN         NaN   66.516070  ...   30.559074   
2      42.201871   70.608908   37.494939   93.652240  ...   47.434076   
3      87.026850  103.104328   95.517008   74.277057  ...   96.743743   
4            NaN         NaN         NaN         NaN  ...         NaN   
...          ...         ...         ...         ...  ...         ...   
4463   68.235804   57.382027   78.488982   70.849244  ...   94.210022   
4464   53.125322   63.776844   77.726507   65.571384  ...   51.040246   
4465   71.231746   68.090875   84.631091   75.376484  ...   71.048504   
4466   58.311372   73.901383   59.236189   60.870573  ...   62.027236   
4467   70.706313         NaN   71.589109         NaN  ...   84.834010   

      2008-06-02  2008-06-03  2008-06-04  2008-06-05  2008-06-06  2008-06-07  \
0      86.778021         NaN         NaN  102.965210         NaN         NaN   
1      13.409085         NaN         NaN         NaN   10.202780         NaN   
2     116.092321   66.100166   38.033621   72.804047   50.743756   73.163894   
3      91.820906   95.574198   86.636611   88.155476   95.542178  100.308692   
4            NaN   47.166686         NaN         NaN         NaN         NaN   
...          ...         ...         ...         ...         ...         ...   
4463   71.199628   63.996375   82.253467   63.133191   74.213952   70.793947   
4464   72.650004   65.053001   67.198314   64.945885   57.927142   61.617524   
4465   66.083079   87.441198   65.877720   52.069321   31.404979   57.313889   
4466   54.819899   62.514630   48.415416   74.165651   64.871825   57.911369   
4467   54.972848   36.596823   29.515865         NaN   85.866007         NaN   

      2008-06-08  2008-06-09  2008-06-10  
0            NaN         NaN         NaN  
1            NaN         NaN         NaN  
2      72.856685   62.130697         NaN  
3      90.068414   93.511852  101.564905  
4            NaN         NaN         NaN  
...          ...         ...         ...  
4463   71.254088   66.452805   60.447080  
4464   59.310603   62.062505         NaN  
4465   68.303598   46.430845   72.351216  
4466   67.838563   64.668299   92.997165  
4467         NaN   26.557968         NaN  

[4468 rows x 25 columns]

In [26]:
# 시간 단위 데이터프레임 생성
time_df = pd.DataFrame()
time_df.index = df[0].index
for i in range(0, len(df.columns), 12):
    cols = df.columns[i:i+12]
    sums = df[cols].mean(axis=1)
    colname = start_date + timedelta(hours=int(i/12))

    time_df = pd.concat((time_df, sums.rename(colname)), axis=1)

time_df

2008-05-17 00:00:00  2008-05-17 01:00:00  2008-05-17 02:00:00  \
0                     NaN                  NaN                  NaN   
1                     NaN                  NaN                  NaN   
2                     NaN                  NaN                  NaN   
3                     NaN                  NaN                  NaN   
4                     NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
4463                  NaN                  NaN                  NaN   
4464                  NaN                  NaN                  NaN   
4465                  NaN                  NaN                  NaN   
4466                  NaN                  NaN                  NaN   
4467                  NaN                  NaN                  NaN   

      2008-05-17 03:00:00  2008-05-17 04:00:00  2008-05-17 05:00:00  \
0                     NaN                  NaN                  NaN   
1                     NaN                  NaN                  NaN   
2                     NaN                  NaN                  NaN   
3                     NaN                  NaN                  NaN   
4                     NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
4463                  NaN                  NaN                  NaN   
4464                  NaN                  NaN                  NaN   
4465                  NaN                  NaN                  NaN   
4466                  NaN                  NaN                  NaN   
4467                  NaN                  NaN                  NaN   

      2008-05-17 06:00:00  2008-05-17 07:00:00  2008-05-17 08:00:00  \
0                     NaN                  NaN                  NaN   
1                     NaN                  NaN                  NaN   
2                     NaN                  NaN                  NaN   
3                     NaN                  NaN                  NaN   
4                     NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
4463                  NaN                  NaN                  NaN   
4464                  NaN                  NaN                  NaN   
4465                  NaN                  NaN                  NaN   
4466                  NaN                  NaN                  NaN   
4467                  NaN                  NaN                  NaN   

      2008-05-17 09:00:00  ...  2008-06-10 00:00:00  2008-06-10 01:00:00  \
0                     NaN  ...                  NaN                  NaN   
1                     NaN  ...                  NaN                  NaN   
2                     NaN  ...                  NaN                  NaN   
3                     NaN  ...                  NaN                  NaN   
4                     NaN  ...                  NaN                  NaN   
...                   ...  ...                  ...                  ...   
4463                  NaN  ...                  NaN                  NaN   
4464                  NaN  ...                  NaN                  NaN   
4465                  NaN  ...                  NaN                  NaN   
4466                  NaN  ...                  NaN                  NaN   
4467                  NaN  ...                  NaN                  NaN   

      2008-06-10 02:00:00  2008-06-10 03:00:00  2008-06-10 04:00:00  \
0                     NaN                  NaN                  NaN   
1                     NaN                  NaN                  NaN   
2                     NaN                  NaN                  NaN   
3              101.564905                  NaN                  NaN   
4                     NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
4463            60.447080            

In [27]:
ffs_date_df = pd.DataFrame(columns=date_df.columns)
for c in ffs_date_df.columns:
    ffs_date_df[c] = ffs_df['0']
tci_date_df = pd.DataFrame(date_df.values / ffs_date_df.values, columns=date_df.columns).transpose()
tci_date_df.index.name = "Date"
tci_date_df

0         1         2         3         4     5     6     \
Date                                                                       
2008-05-17  2.301809  0.274829  1.605536  3.611331       NaN   NaN   NaN   
2008-05-18  2.444643       NaN  1.645234  2.568088       NaN   NaN   NaN   
2008-05-19       NaN  1.920555  2.853592  3.202670       NaN   NaN   NaN   
2008-05-20       NaN       NaN  2.022833  3.059881       NaN   NaN   NaN   
2008-05-21       NaN       NaN  2.036664  3.162641       NaN   NaN   NaN   
2008-05-22  2.153674  0.590066  2.590380  3.128039       NaN   NaN   NaN   
2008-05-23  1.339064       NaN  1.387763  3.338199       NaN   NaN   NaN   
2008-05-24       NaN       NaN  2.321898  3.954903       NaN   NaN   NaN   
2008-05-25       NaN       NaN  1.232981  3.663867       NaN   NaN   NaN   
2008-05-26       NaN  2.187309  3.079653  2.849139       NaN   NaN   NaN   
2008-05-27       NaN       NaN  2.100132  3.609474       NaN   NaN   NaN   
2008-05-28       NaN  3.432306  2.114554  2.682681       NaN   NaN   NaN   
2008-05-29  2.116910       NaN  0.524643  3.395513       NaN   NaN   NaN   
2008-05-30  1.490919       NaN  2.275679  3.163367       NaN   NaN   NaN   
2008-05-31       NaN       NaN  3.287955  3.384363       NaN   NaN   NaN   
2008-06-01       NaN  1.004902  1.559818  3.710922       NaN   NaN   NaN   
2008-06-02  1.997652  0.440943  3.817571  3.522091       NaN   NaN   NaN   
2008-06-03       NaN       NaN  2.173633  3.666061  1.551026   NaN   NaN   
2008-06-04       NaN       NaN  1.250695  3.323230       NaN   NaN   NaN   
2008-06-05  2.370286       NaN  2.394082  3.381491       NaN   NaN   NaN   
2008-06-06       NaN  0.335507  1.668654  3.664832       NaN   NaN   NaN   
2008-06-07       NaN       NaN  2.405916  3.847668       NaN   NaN   NaN   
2008-06-08       NaN       NaN  2.395813  3.454868       NaN   NaN   NaN   
2008-06-09       NaN       NaN  2.043101  3.586953       NaN   NaN   NaN   
2008-06-10       NaN       NaN       NaN  3.895854       NaN   NaN   NaN   

            7         8         9     ...      4458      4459      4460  \
Date                                  ...                                 
2008-05-17   NaN       NaN       NaN  ...  3.275878  2.767461       NaN   
2008-05-18   NaN       NaN  3.730788  ...  2.929689  2.053240  1.526552   
2008-05-19   NaN       NaN  2.834058  ...       NaN  2.343531  2.758569   
2008-05-20   NaN       NaN       NaN  ...       NaN  2.033488  0.915399   
2008-05-21   NaN       NaN  1.886729  ...       NaN  2.209878  2.736161   
2008-05-22   NaN       NaN  2.081974  ...  0.930540  2.394731  1.446595   
2008-05-23   NaN       NaN       NaN  ...  1.922212  2.292448  1.816958   
2008-05-24   NaN       NaN       NaN  ...  2.253537  2.167174  1.948406   
2008-05-25   NaN       NaN       NaN  ...  2.545812  2.441165  1.921748   
2008-05-26   NaN       NaN       NaN  ...  2.932098  2.396835       NaN   
2008-05-27   NaN       NaN  0.401330  ...       NaN  2.415720       NaN   
2008-05-28   NaN       NaN  2.761224  ...  3.543359  2.121892  2.545209   
2008-05-29   NaN       NaN  2.607077  ...  1.080146  2.294652  1.806169   
2008-05-30   NaN       NaN  1.403308  ...       NaN  1.680403  1.784326   
2008-05-31   NaN       NaN  2.085617  ...  2.288270  2.349086  1.430345   
2008-06-01   NaN       NaN       NaN  ...       NaN  2.246957  0.406489   
2008-06-02   NaN       NaN  0.276593  ...  2.473281  2.240570  0.861695   
2008-06-03   NaN       NaN       NaN  ...       NaN  2.336920  1.907949   
2008-06-04   NaN       NaN  2.377253  ...  1.174369  2.667836  2.276971   
2008-06-05   NaN  3.055766  0.503379  ...       NaN  2.235436  1.109327   
2008-06-06   NaN       NaN  1.433502  ...  0.240811  2.534496  2.173903   
2008-06-07   NaN       NaN  1.831306  ...  2.352918  2.378771  1.582821   
2008-06-08   NaN       NaN       NaN  ...  1.951724  2.573975  1.842153   
2008-06-09   NaN  0.636643  1.400328  ...       NaN  2.164802  2.737595   
2008-06-10   

In [28]:
ffs_time_df = pd.DataFrame(columns=time_df.columns)
for c in ffs_time_df.columns:
    ffs_time_df[c] = ffs_df['0']
tci_time_df = pd.DataFrame(time_df.values / ffs_time_df.values, columns=time_df.columns).transpose()
tci_time_df.index.name = "Time"
tci_time_df

0     1     2     3     4     5     6     7     8     \
Time                                                                        
2008-05-17 00:00:00   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2008-05-17 01:00:00   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2008-05-17 02:00:00   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2008-05-17 03:00:00   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2008-05-17 04:00:00   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
...                   ...   ...   ...   ...   ...   ...   ...   ...   ...   
2008-06-10 05:00:00   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2008-06-10 06:00:00   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2008-06-10 07:00:00   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2008-06-10 08:00:00   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2008-06-10 09:00:00   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

                     9     ...  4458      4459  4460  4461  4462  4463  4464  \
Time                       ...                                                 
2008-05-17 00:00:00   NaN  ...   NaN       NaN   NaN   NaN   NaN   NaN   NaN   
2008-05-17 01:00:00   NaN  ...   NaN       NaN   NaN   NaN   NaN   NaN   NaN   
2008-05-17 02:00:00   NaN  ...   NaN       NaN   NaN   NaN   NaN   NaN   NaN   
2008-05-17 03:00:00   NaN  ...   NaN       NaN   NaN   NaN   NaN   NaN   NaN   
2008-05-17 04:00:00   NaN  ...   NaN       NaN   NaN   NaN   NaN   NaN   NaN   
...                   ...  ...   ...       ...   ...   ...   ...   ...   ...   
2008-06-10 05:00:00   NaN  ...   NaN       NaN   NaN   NaN   NaN   NaN   NaN   
2008-06-10 06:00:00   NaN  ...   NaN  2.097533   NaN   NaN   NaN   NaN   NaN   
2008-06-10 07:00:00   NaN  ...   NaN       NaN   NaN   NaN   NaN   NaN   NaN   
2008-06-10 08:00:00   NaN  ...   NaN       NaN   NaN   NaN   NaN   NaN   NaN   
2008-06-10 09:00:00   NaN  ...   NaN       NaN   NaN   NaN   NaN   NaN   NaN   

                     4465  4466  4467  
Time                                   
2008-05-17 00:00:00   NaN   NaN   NaN  
2008-05-17 01:00:00   NaN   NaN   NaN  
2008-05-17 02:00:00   NaN   NaN   NaN  
2008-05-17 03:00:00   NaN   NaN   NaN  
2008-05-17 04:00:00   NaN   NaN   NaN  
...                   ...   ...   ...  
2008-06-10 05:00:00   NaN   NaN   NaN  
2008-06-10 06:00:00   NaN   NaN   NaN  
2008-06-10 07:00:00   NaN   NaN   NaN  
2008-06-10 08:00:00   NaN   NaN   NaN  
2008-06-10 09:00:00   NaN   NaN   NaN  

[586 rows x 4468 columns]

In [29]:
tci_date_df.to_csv("../static/data/sanfrancisco/road_tci_date.csv")
tci_time_df.to_csv("../static/data/sanfrancisco/road_tci_time.csv")

# 전체 TCI

In [30]:
city_tci_date_df = pd.DataFrame(tci_date_df.mean(axis=1), columns=["Actual"])
city_tci_date_df

Actual
Date                
2008-05-17  2.406282
2008-05-18  2.387212
2008-05-19  2.407215
2008-05-20  2.421994
2008-05-21  2.430569
2008-05-22  2.392122
2008-05-23  2.398785
2008-05-24  2.412755
2008-05-25  2.435264
2008-05-26  2.473952
2008-05-27  2.414354
2008-05-28  2.406573
2008-05-29  2.396288
2008-05-30  2.365882
2008-05-31  2.405246
2008-06-01  2.434292
2008-06-02  2.404433
2008-06-03  2.407813
2008-06-04  2.398035
2008-06-05  2.394371
2008-06-06  2.377926
2008-06-07  2.429456
2008-06-08  2.446129
2008-06-09  2.449048
2008-06-10  2.578431

In [36]:
city_tci_time_df = pd.DataFrame(tci_time_df.mean(axis=1), columns=["Actual"])
city_tci_time_df

Actual
Time                         
2008-05-17 00:00:00       NaN
2008-05-17 01:00:00       NaN
2008-05-17 02:00:00       NaN
2008-05-17 03:00:00       NaN
2008-05-17 04:00:00       NaN
...                       ...
2008-06-10 05:00:00  2.631505
2008-06-10 06:00:00  2.809083
2008-06-10 07:00:00  2.845597
2008-06-10 08:00:00  2.728040
2008-06-10 09:00:00  2.793357

[586 rows x 1 columns]

In [32]:
city_tci_date_df.to_csv("../static/data/sanfrancisco/city_tci_date.csv")
city_tci_time_df.to_csv("../static/data/sanfrancisco/city_tci_time.csv")